src/Twig/Extension/SonataAdminExtension.php

```diff
@@ -50,8 +54,17 @@ class SonataAdminExtension extends AbstractExtension
      */
     private $xEditableTypeMapping = [];
 
-    public function __construct(Pool $pool, LoggerInterface $logger = null, TranslatorInterface $translator = null)
-    {
+    /**
+     * @var ContainerInterface
+     */
+    private $templateRegistries;
+
+    public function __construct(
+        Pool $pool,
+        LoggerInterface $logger = null,
+        TranslatorInterface $translator = null,
+        ContainerInterface $templateRegistries = null
```

Comment:
is this type hint correct?

Reply:
For now, I'll replace it with a ServiceLocator in a follow-up RP.

Reply:
The service locator implements this interface, right?

Reply:
No, it implements `Psr\Container\ContainerInterface` instead of `Symfony\Component\DependencyInjection\ContainerInterface`.

Reply:
Fortunately, this is not confusing at all. ;-)

Reply:
> When we drop support for <3.3, I can replace most references to Symfony\Component\DependencyInjection\ContainerInterface with Psr\Container\ContainerInterface.

And that will be a BC break :P

Reply:
Indeed. But that's a problem for another time.

Reply:
But we said we will do this on stable as we don't have to support 2 LTS version (2.8 and 3.4)?

Reply:
I think we did say that. So that would make it not a problem for another time, but a problem for now.

Reply:
So I should implement the Service Locator in this PR?

Reply:
You can drop a version support, it's BC. The only rule is to drop it if the BC tricks are too heavy. Please see: https://github.com/sonata-project/SonataAdminBundle/blob/3.x/CONTRIBUTING.md#dependency-changes

Also, if BC tricks are in place concerning old dependency support, they should be removed as well.

Two options: On a separate commit or in a separate PR. The separate PR would be preferable because this one is already quite big.

Reply:
I also prefer a separate PR. Thank you for the extra info.

Reply:
Maybe `$templateRegistryLocator` would be a better name to use with the `ContainerInterface`, but it's so long. :-(

Reply:
It is descriptive? I don't mind long name, not sure what others think.

Reply:
I was quite fine with `templateRegistries`, the fact that it's a container is kind of an implementation detail